tobler/util/util.py

```diff
@@ -102,14 +106,76 @@ def project_gdf(gdf, to_crs=None, to_latlong=False):
 
     # calculate the centroid of the union of all the geometries in the
     # GeoDataFrame
-    avg_longitude = gdf['geometry'].unary_union.centroid.x
+    avg_longitude = gdf["geometry"].unary_union.centroid.x
 
     # calculate the UTM zone from this avg longitude and define the UTM
     # CRS to project
-    utm_zone = int(math.floor((avg_longitude + 180) / 6.) + 1)
-    utm_crs = '+proj=utm +zone={} +ellps=WGS84 +datum=WGS84 +units=m +no_defs'.format(utm_zone)
+    utm_zone = int(math.floor((avg_longitude + 180) / 6.0) + 1)
+    utm_crs = "+proj=utm +zone={} +ellps=WGS84 +datum=WGS84 +units=m +no_defs".format(
+        utm_zone
+    )
 
     # project the GeoDataFrame to the UTM CRS
     projected_gdf = gdf.to_crs(utm_crs)
 
     return projected_gdf
+
+
+def hexify(source, resolution=6, clip=False):
+    """Generate a hexgrid geodataframe that covers the face of a source geodataframe.
+
+    Parameters
+    ----------
+    source : geopandas.GeoDataFrame
+        GeoDataFrame to transform into a hexagonal grid
+    resolution : int, optional
+        resolution of output h3 hexgrid. 
+        See <https://h3geo.org/docs/core-library/restable> for more information
+    clip : bool, optional
+        if True, hexagons are clipped to the precise boundary of the source gdf. Otherwise,
+        heaxgons along the boundary will be left intact.
+
+    Returns
+    -------
+    geopandas.GeoDataFrame
+        a GeoDataFrame whose rows comprise a hexagonal h3 grid. 
+    """
+    try:
+        from h3 import h3
+    except ImportError:
+        raise ImportError(
+            "This function requires the `h3` library. "
+            "You can install it with `conda install h3` or "
+            "`pip install h3`"
+        )
+    orig_crs = source.crs.to_string()
+
+    if not source.crs.name == "WGS 84":
+        source = source.copy().to_crs(4326)
+
+    hexids = pandas.Series(
+        list(
+            h3.polyfill(
+                source.unary_union.__geo_interface__,
+                resolution,
+                geo_json_conformant=True,
+            )
+        ),
+        name="hex_id",
+    )
+
+    polys = hexids.apply(
```

Comment:
This could potentially be processed in parallel for faster execution in modern hardware. Chunk `hexids`, send to each core (similar to how we're thinking about it in #112) and collect later. It should scale _almost_ linearly.

Reply:
good idea. I wonder if we could also outsource some of that logic to something like [`modin`](https://modin.readthedocs.io/en/latest/)?

Reply:
> I wonder if we could also outsource some of that logic to something

That should be ideally `dask-geopandas` which is under development.

Reply:
+1, though in this case the operation is just an `apply`, which shouldn't need dask-geopandas (and as far as i understand [dask doesnt do apply](https://modin.readthedocs.io/en/latest/comparisons/dask.html)?)

so in a perfect world, maybe dask-geopandas could end up as an additional backend for modin or something, depending on the operation?

Reply:
Dask does row-wise `apply` very efficiently. It cannot do column-wise `apply` but that is not an issue here.

It is true that we can have `hexids` as `dask.Series` and then apply remains the same. Then below we'll just call `compute()` before creating the GeoDataFrame. That should be relatively straightforward to implement with a vanilla `dask.DataFrame` even now.

Reply:
sweet. would the idea be to check whether dask in installed, and then use multicore handling and fallback if not, or to make dask a full dep?

Reply:
I would add a keyword to control that, similar to this implementation we have in momepy: https://github.com/martinfleis/momepy/blob/19e3a1a6eb9577a2c160710bd139034a459a4f97/momepy/elements.py#L433.

There may be cases when single-core is faster than dask (typically small areas) due to to the scheduler overhead.

Reply:
cool, thanks for the tip. i'm gonna go ahead and merge this implementation so we all have it available, then we can hack on the multicore enhancement